src/renderer/vsync/vsync_macos.rs

```diff
@@ -0,0 +1,76 @@
+use std::{ffi::c_void, sync::Barrier};
+
+use crate::profiling::tracy_zone;
+
+use crate::renderer::WindowedContext;
+
+use super::macos_display_link::{core_video, get_display_id_of_window, MacosDisplayLink};
+
+pub struct VSyncMacos {
+    old_display: core_video::CGDirectDisplayID,
+    display_link: Option<MacosDisplayLink>,
+    barrier: Barrier,
+}
+
+#[allow(unused_variables, non_snake_case)]
+extern "C" fn callback(
+    displayLink: core_video::CVDisplayLinkRef,
+    inNow: *const core_video::CVTimeStamp,
+    inOutputTime: *const core_video::CVTimeStamp,
+    flagsIn: u64,
+    flagsOut: *mut u64,
+    displayLinkContext: *mut c_void,
+) -> core_video::CVReturn {
+    tracy_zone!("VSyncDisplayLinkCallback");
+
+    // The display link should be dropped before vsync, so this should be safe.
+    let vsync = displayLinkContext as *mut VSyncMacos;
+
+    unsafe {
+        (*vsync).barrier.wait(); // TODO: Maybe there is a safer way.
```

Comment:
I think it would be better to use a synchronisation primitive that does not block the callback thread. For example a condition variable like the windows implementation. 

That also makes it possible to wait for vsync with a timeout in the wait function. I think that might be needed to to allow the main thread to wake up when there are problems, like the display link getting invalid because of a monitor disconnect or something.

Reply:
Ok. I'll take a look this night. The display link callback is called in a separate thread according to apple's document. So the implementation should be quite similar.

Reply:
@fredizzimo Do you think it is still necessary to use `Arc` here like on Windows? The display link thread should be destroyed once the vsync object is dropped. But I'm not sure. Apple's documentation is very unclear.

Reply:
I think the arc was used just to be able to be able to share the same Mutex and Condvar  and in the update loop. So I believe it's still required.

Reply:
Hm, now that I think about it, you probably don't need the arc. I think it's enough to store the objects directly on self, like you did for the barrier.

Reply:
I think the arc used on windows is to avoid that lifetime of vsync thread does not lie in lifetime of main thread. So arc can avoid use-after-release. On macos, the issue is the same. We can't promise the lifetime relationship of two threads.
However, the barrier should be wrong. Because if the display link fails, main thread will block forever since the barrier never counts down to 0. Actually I have marked this in code at `wait_for_vsync`.

Reply:
I think a little bit more. If I need to share sync primitives between two threads, use arc is the correct way in rust. This unsafe cast of vsync just breaks the semantic of rust though it might work well.

Reply:
Yes, I'm too inexperienced with rust to give any direct recommendations. But you probably need to create a struct containing the arcs, cast that and send it to the callback.

Reply:
That's what I think. I'll update the code tomorrow. Now I just mark this as resolved.